src/machine/machine_esp32.go

```diff
@@ -20,18 +20,43 @@ type PinMode uint8
 const (
 	PinOutput PinMode = iota
 	PinInput
+	PinInputPullup
+	PinInputPulldown
 )
 
 // Configure this pin with the given configuration.
 func (p Pin) Configure(config PinConfig) {
-	if config.Mode == PinOutput {
+	var muxConfig uint32 // The mux configuration.
```

Comment:
Perhaps all this should be broken out into a separate function, since as you mention a variation of setting each of these will be required for I2C.

Reply:
I'll try implementing SPI to see whether this abstraction works.

Reply:
I implemented SPI and it requires only a small modification to this function.